src/sinks/util/http.rs

```diff
@@ -598,6 +599,144 @@ pub fn validate_headers(
     Ok(validated_headers)
 }
 
+/// Request type for use in the `Service` implementation of HTTP stream sinks.
+#[derive(Clone)]
+pub struct HttpRequest {
+    payload: Bytes,
+    finalizers: EventFinalizers,
+    request_metadata: RequestMetadata,
+}
+
+impl HttpRequest {
+    /// Creates a new `HttpRequest`.
+    pub fn new(
+        payload: Bytes,
+        finalizers: EventFinalizers,
+        request_metadata: RequestMetadata,
+    ) -> Self {
+        Self {
+            payload,
+            finalizers,
+            request_metadata,
+        }
+    }
+}
+
+impl Finalizable for HttpRequest {
+    fn take_finalizers(&mut self) -> EventFinalizers {
+        self.finalizers.take_finalizers()
+    }
+}
+
+impl MetaDescriptive for HttpRequest {
+    fn get_metadata(&self) -> &RequestMetadata {
+        &self.request_metadata
+    }
+
+    fn metadata_mut(&mut self) -> &mut RequestMetadata {
+        &mut self.request_metadata
+    }
+}
+
+impl ByteSizeOf for HttpRequest {
+    fn allocated_bytes(&self) -> usize {
+        self.payload.allocated_bytes() + self.finalizers.allocated_bytes()
+    }
+}
+
+/// Response type for use in the `Service` implementation of HTTP stream sinks.
+pub struct HttpResponse {
+    pub http_response: Response<Bytes>,
+    events_byte_size: GroupedCountByteSize,
+    raw_byte_size: usize,
+}
+
+impl DriverResponse for HttpResponse {
+    fn event_status(&self) -> EventStatus {
+        if self.http_response.status().is_success() {
+            EventStatus::Delivered
+        } else {
+            EventStatus::Rejected
+        }
+    }
+
+    fn events_sent(&self) -> &GroupedCountByteSize {
+        &self.events_byte_size
+    }
+
+    fn bytes_sent(&self) -> Option<usize> {
+        Some(self.raw_byte_size)
+    }
+}
+
+/// HTTP request builder for batched HTTP stream sinks using the generic `HttpService`
+pub trait HttpServiceRequestBuilder {
+    fn build(&self, body: Bytes) -> Request<Bytes>;
+}
+
+/// Generic 'Service' implementation for batched HTTP stream sinks.
+#[derive(Clone)]
+pub struct HttpService<B> {
+    batch_service:
+        HttpBatchService<BoxFuture<'static, Result<Request<Bytes>, crate::Error>>, HttpRequest>,
+    _phantom: PhantomData<B>,
+}
+
+impl<B> HttpService<B>
+where
+    B: HttpServiceRequestBuilder + std::marker::Sync + std::marker::Send + 'static,
+{
+    pub fn new(http_client: HttpClient<Body>, http_request_builder: B) -> Self {
+        let http_request_builder = Arc::new(http_request_builder);
+
+        let batch_service = HttpBatchService::new(http_client, move |req| {
+            let req: HttpRequest = req;
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏

```suggestion
        let batch_service = HttpBatchService::new(http_client, move |req: HttpRequest| {
```

Just a little nicer I think.